pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,321 @@
+        return axis_vars[kernel.axis]*poly
+    elif isinstance(kernel, AxisSourceDerivative):
+        poly = _convert_kernel_to_poly(kernel.inner_kernel, axis_vars)
+        return -axis_vars[kernel.axis]*poly
+    return 1
+
+
+def _convert_source_poly_to_int_g(poly, orig_int_g, axis_vars):
+    """This does the opposite of :func:`_convert_kernel_to_poly`
+    and converts a polynomial back to a source derivative
+    operator. First it is converted to a :class:`sumpy.kernel.Kernel`
+    and then to a :class:`~pytential.symbolic.primitives.IntG`.
+    """
+    from pytential.symbolic.pde.system_utils import simplify_densities
+    to_pymbolic = SympyToPymbolicMapper()
+
+    orig_kernel = orig_int_g.source_kernels[0]
+    source_kernels = []
+    densities = []
+    for monom, coeff in poly.terms():
+        kernel = orig_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisSourceDerivative(idim, kernel)
+        source_kernels.append(kernel)
+        densities.append(to_pymbolic(coeff) * (-1)**sum(monom))
+    return orig_int_g.copy(source_kernels=tuple(source_kernels),
+            densities=tuple(simplify_densities(densities)))
+
+
+def _convert_target_poly_to_int_g(poly, orig_int_g, rhs_int_g):
+    """This does the opposite of :func:`_convert_kernel_to_poly`
+    and converts a polynomial back to a target derivative
+    operator. It is applied to a :class:`~pytential.symbolic.primitives.IntG`
+    object and returns a new instance.
+    """
+    to_pymbolic = SympyToPymbolicMapper()
+
+    result = 0
+    for monom, coeff in poly.terms():
+        kernel = orig_int_g.target_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisTargetDerivative(idim, kernel)
+        result += orig_int_g.copy(target_kernel=kernel,
+                source_kernels=rhs_int_g.source_kernels,
+                densities=rhs_int_g.densities) * to_pymbolic(coeff)
+
+    return result
+
+
+class CoefficientCollector(CoefficientCollectorBase):
+    """This extends :class:`pymbolic.mapper.coefficient.CoefficientCollector`
+    by extending the targets to be any expression instead of just algebraic
+    leafs
+    """
+    def __init__(self, source_dependent_variables):
+        self.source_dependent_variables = source_dependent_variables
+
+    def __call__(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        return super().__call__(expr)
+
+    def map_algebraic_leaf(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        else:
+            return {1: expr}
+
+    rec = __call__
+
+
+def _syzygy_module(m, gens):
+    """Takes as input a module of polynomials with domain `sympy.EX` represented
+    as a matrix and returns the syzygy module as a matrix of polynomials in the
+    same domain.
```

Comment:
Describe the property satisfied by the returned syzygy modules.

Reply:
Done in 4ed556e70d5221e1458aa794e94802688798a33a